thrust/thrust/system/detail/generic/select_system.h

```diff
@@ -26,90 +26,66 @@
 #elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
 #  pragma system_header
 #endif // no system header
+
 #include <thrust/detail/execution_policy.h>
 #include <thrust/detail/type_traits.h>
 #include <thrust/iterator/detail/any_system_tag.h>
 #include <thrust/iterator/detail/device_system_tag.h>
 #include <thrust/iterator/detail/minimum_system.h>
 
 THRUST_NAMESPACE_BEGIN
-namespace system
-{
-namespace detail
-{
-namespace generic
-{
-
-template <typename Tag>
-struct select_system1_exists;
-
-template <typename Tag1, typename Tag2>
-struct select_system2_exists;
 
-template <typename Tag1, typename Tag2, typename Tag3>
-struct select_system3_exists;
+// forward declaration of any_system_tag for any_conversion below
+struct any_system_tag;
 
-template <typename Tag1, typename Tag2, typename Tag3, typename Tag4>
-struct select_system4_exists;
-
-template <typename Tag1, typename Tag2, typename Tag3, typename Tag4, typename Tag5>
-struct select_system5_exists;
+namespace system::detail::generic
+{
+template <typename, typename... Tags>
+inline constexpr bool select_system_exists = false;
 
-template <typename Tag1, typename Tag2, typename Tag3, typename Tag4, typename Tag5, typename Tag6>
-struct select_system6_exists;
+template <typename... Tags>
+inline constexpr bool
+  select_system_exists<::cuda::std::void_t<decltype(select_system(::cuda::std::declval<Tags>()...))>, Tags...> = true;
```

Comment:
Shouldnt this check for `minimum_type` not `select_system`

Reply:
My interpretation of the old implementation in `thrust/system/detail/generic/select_system_exists.h` was that it tries to check whether `select_system(Tags...)` resolves to a function or not. Why do you think it should be `minimum_type`?

Reply:
Because that is what is actually checked inside the implementation of `select_system`

Reply:
IIUC `select_system` can be provided by users for their own systems. I not (yet) confident enough to refactor beyond the previous behavior.

Reply:
I have split the `select_system` refactoring out into #4762 to figure out what's bothering MSVC.